src/httpd.c

```diff
@@ -389,65 +389,57 @@ serve_file(struct evhttp_request *req, const char *uri)
       return;
     }
 
-  ret = lstat(path, &sb);
-  if (ret < 0)
+
+  if (!realpath(path, deref))
     {
-      DPRINTF(E_WARN, L_HTTPD, "Could not lstat() %s: %s\n", path, strerror(errno));
+      DPRINTF(E_LOG, L_HTTPD, "Could not dereference %s: %s\n", path, strerror(errno));
 
       httpd_send_error(req, HTTP_NOTFOUND, "Not Found");
-
       return;
     }
 
-  if (S_ISLNK(sb.st_mode))
+  if (strlen(deref) >= PATH_MAX)
     {
-      if (!realpath(path, deref))
-	{
-	  DPRINTF(E_LOG, L_HTTPD, "Could not dereference %s: %s\n", path, strerror(errno));
-
-	  httpd_send_error(req, HTTP_NOTFOUND, "Not Found");
-
-	  return;
-	}
+      DPRINTF(E_LOG, L_HTTPD, "Dereferenced path exceeds PATH_MAX: %s\n", path);
 
-      if (strlen(deref) + 1 > PATH_MAX)
-	{
-	  DPRINTF(E_LOG, L_HTTPD, "Dereferenced path exceeds PATH_MAX: %s\n", path);
-
-	  httpd_send_error(req, HTTP_NOTFOUND, "Not Found");
-
-	  return;
-	}
+      httpd_send_error(req, HTTP_NOTFOUND, "Not Found");
 
-      strcpy(path, deref);
+      return;
+    }
 
-      ret = stat(path, &sb);
-      if (ret < 0)
-	{
-	  DPRINTF(E_LOG, L_HTTPD, "Could not stat() %s: %s\n", path, strerror(errno));
+  ret = lstat(deref, &sb);
+  if (ret < 0)
+    {
+      DPRINTF(E_WARN, L_HTTPD, "Could not lstat() %s: %s\n", deref, strerror(errno));
 
-	  httpd_send_error(req, HTTP_NOTFOUND, "Not Found");
+      httpd_send_error(req, HTTP_NOTFOUND, "Not Found");
 
-	  return;
-	}
+      return;
     }
 
   if (S_ISDIR(sb.st_mode))
     {
       slashed = (path[strlen(path) - 1] == '/');
+      strncat(path, ((slashed) ? "index.html" : "/index.html"), sizeof(path) - strlen(path));
```

Comment:
A bit "unclean" that path and deref are used interchangebly in the function? path is also addressed again below in line 502. Perhaps serve_file() should actually be split up a bit.